phylokit/maximum_likelihood/felsenstein.py

```diff
@@ -0,0 +1,275 @@
+    sample_nodes,
+    variant_allele,
+    node_branch_length,
+    rate,
+    pi,
+):
+    ret = np.zeros(call_genotype.shape[0], dtype=np.float64)
+
+    sample_mask = np.zeros(num_nodes, dtype=np.bool_)
+    sample_mask[sample_nodes] = True
+
+    for i in prange(call_genotype.shape[0]):
+        likelihood = np.zeros((num_nodes, 4), dtype=np.float64)
+        for j, sample_node in enumerate(sample_nodes):
+            if call_genotype[i, j] == -1:
+                likelihood[sample_node] = 0.25
+            else:
+                likelihood[sample_node, variant_allele[i, call_genotype[i, j, 0]]] = 1.0
+        for node in traversal_postorder:
+            if sample_mask[node]:
+                continue
+            else:
+                u = node_left_child[node]
+                probs = _calculate_likelihood(
+                    u,
+                    node_branch_length,
+                    rate,
+                    pi,
+                    likelihood,
+                )
+                v = node_right_sib[u]
+                while v != -1:
+                    probs *= _calculate_likelihood(
+                        v,
+                        node_branch_length,
+                        rate,
+                        pi,
+                        likelihood,
+                    )
+                    v = node_right_sib[v]
+
+                likelihood[node] = probs
+
+        ret[i] = np.sum(likelihood[traversal_postorder[-1]] * 0.25)
+
+    return np.prod(ret)
+
+
+def likelihood_felsenstein(
+    ds,
+    rate,
+    GENOTYPE_ARRAY=None,
+    pi=None,
+):
+    """
+    Calculate the log likelihood of a given tree with `Felsenstein`
+    pruning algorithm. This implementation is a parallelized
+    version of the `naive_likelihood_felsenstein` function.
+
+    :param ds: phylokit.DataSet, tree data
+    :param rate: float, mutation rate
+    :param GENOTYPE_ARRAY: list, genotype mapping,
+    default is [b"A", b"C", b"G", b"T"]
+    :param pi: np.array, transtion probability matrix with
+    shape (4, 4),
+    default is [0.25, 0.25, 0.25, 0.25]
+    :return: float, likelihood
+    """
+
+    if GENOTYPE_ARRAY is None:
+        GENOTYPE_ARRAY = np.array(
+            [b"A", b"C", b"G", b"T"],
+            dtype="S",
+        )
+    else:
+        GENOTYPE_ARRAY = np.array(GENOTYPE_ARRAY, dtype="S")
+
+    if pi is None:
+        pi = np.full((4, 4), 0.25, dtype=np.float64)
+    elif pi.shape != (4, 4):
+        raise ValueError("The transition probability matrix must have shape (4, 4)")
+    else:
+        pi = np.array(pi, dtype=np.float64)
+
+    ret = _likelihood_felsenstein(
+        ds.node_left_child.data,
+        ds.node_right_sib.data,
+        ds.call_genotype.data,
+        ds.nodes.shape[0],
+        ds.traversal_postorder.data,
+        ds.sample_node.data,
+        util.base_mapping(ds.variant_allele.data, GENOTYPE_ARRAY),
+        ds.node_branch_length.data,
+        rate,
+        pi,
+    )
+
+    return np.log(ret)
```

Comment:
Taking the log at the end like this will probably result in underflow issues. I think you probably need to do the full calculation in log-space. We can do that in a follow-up though.

Reply:
Removed `np.log`